### PR TITLE
[8.0] mrp_operations_project

### PR DESCRIPTION
**Pull request by pedrobaeza**
# 
# MRP Project Link (with operations)

This module links manufacturing operations with projects. It performs the
following actions:
- Auto-create tasks for work orders with assigned operators.
- Add a tab where coding task works for the work order.
- Add a link on task for the scheduled products for that work order.
# Installation

This modules is auto-installed when you install _mrp_operations_extension_ and
_mrp_project_.
# Usage

In a manufacturing order (MO) that has work orders, and they have at least
one operator assigned, when you start that work order, a task is created with
that operator as responsible of it.

Navigating through each work order, there's a new tab called "Operators time"
where the task works can be put in an easy way.

And last, if you go to a task linked to a work order, you'll see the scheduled
products for that work order.

### CODE REVIEW

**Comment by pedrobaeza** [comment]
This depends on https://github.com/OCA/manufacture/pull/57
==========================================================

The only relevant commit is https://github.com/pedrobaeza/manufacture/commit/b543da32eae75d19e4ac1e5868cc6538d657c0b9

But this way you can meanwhile check this one functionally.

cc @rafaelbn @anajuaristi @javieriniesta @JMODaERP

**Comment by anajuaristi** [comment]
:+1: From my side

Tested:
1. Modify a routing to include 3 routing lines
2. Modify one routing line to include several machines. Change the one by default
3. Include 0, 1 and 2 workers on each default routing line
4. Include 2 cycles/1 hour, 1cycle/1hour, 1cycle/1.5hours on each default routing line
5. Launch a manufacturing order

Verified and correct: 
Calculation of cycle and hours in work orders using default machines
Default assigned machines on each work order
Task creation starting one operation

File: mrp_hook/README.rst [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+===============================
+Hooks for enabling advanced MRP
+===============================
+
+Technical module that provides the proper framework infrastructure (hooks,
+fallback, etc) to enable advanced functionality in the manufacturing area,
+as https://github.com/odoo/odoo/pull/8885 hasn't been accepted for v8.
```
Comment (elicoidal) [inline mrp_hook/README.rst]: colon?

File: mrp_hook/README.rst [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+* Hooks in *_bom_explode* method for returning dictionary for consume and
```
Comment (elicoidal) [inline mrp_hook/README.rst]: "to return a dictionary for consumption and"

File: mrp_hook/models/mrp_bom.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
+# -*- coding: utf-8 -*-
+# (c) 2015 Serv. Tecnol. Avanzados - Pedro M. Baeza
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+
+from openerp import models, api, tools, _
```
Comment (elicoidal) [inline mrp_hook/models/mrp_bom.py]: Please sort in alphabetical order following: https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#imports

File: mrp_operations_extension/README.rst [hunk 4]
```diff
@@ -0,0 +1,76 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+===============================
+MRP operations extension module
+===============================
+
+This module adds:
```
Comment (elicoidal) [inline mrp_operations_extension/README.rst:9]: @pedrobaeza module naming is generic (I never like blablable_extension modules) and I have the feeling that many different concepts are added in one module. 
Would  it be possible to split them in different more specific modules?

File: mrp_operations_extension/__init__.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
+# -*- coding: utf-8 -*-
```
Comment (elicoidal) [inline mrp_operations_extension/__init__.py]: Please use simple headers following OCA templates
https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py#L3

File: mrp_operations_extension/models/__init__.py [hunk 6]
```diff
@@ -0,0 +1,11 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+# For copyright and license notices, see __openerp__.py file in root directory
+##############################################################################
+from . import mrp_routing
```
Comment (elicoidal) [inline mrp_operations_extension/models/__init__.py]: Please sort in alphabetical order following: https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#imports

File: mrp_operations_extension/models/mrp_bom.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+    @api.multi
+    @api.onchange('routing_id')
+    def onchange_routing_id(self):
+        for line in self.bom_line_ids:
+            line.operation = self.routing_id.workcenter_lines[:1]
+        if self.routing_id:
+            return {'warning': {
+                    'title': _('Changing Routing'),
+                    'message': _("Changing routing will cause to change the"
```
Comment (elicoidal) [inline mrp_operations_extension/models/mrp_bom.py]: "Changing routing will change the operation of each consumed component . By default the first one in the routing will be set by default."

File: mrp_operations_extension/models/mrp_routing.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+                break
+
+
+class MrpOperationWorkcenter(models.Model):
+    _name = 'mrp.operation.workcenter'
+    _description = 'MRP Operation Workcenter'
+
+    custom_data = fields.Boolean(
+        string="Custom", default=False,
+        help="If you mark this check, this means that the work center in this "
```
Comment (elicoidal) [inline mrp_operations_extension/models/mrp_routing.py:87]: "If checked, the work center in this routing will have a different capacity data than the one defined in the work center itself."

File: mrp_operations_extension/models/mrp_routing.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+             "routing has different capacity data than the defined on the "
+             "work center itself")
+    workcenter = fields.Many2one(
+        'mrp.workcenter', string='Workcenter', required=True)
+    routing_workcenter = fields.Many2one(
+        'mrp.routing.workcenter', 'Routing workcenter', required=True)
+    time_efficiency = fields.Float('Efficiency factor')
+    capacity_per_cycle = fields.Float('Capacity per cycle')
+    time_cycle = fields.Float('Time for 1 cycle (hours)',
+                              help="Time in hours for doing one cycle.")
```
Comment (elicoidal) [inline mrp_operations_extension/models/mrp_routing.py]: "Duration for one cycle"

File: mrp_operations_extension/models/mrp_routing.py [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    time_start = fields.Float('Time before prod.',
+                              help="Time in hours for the setup.")
```
Comment (elicoidal) [inline mrp_operations_extension/models/mrp_routing.py]: Duration necessary before the cycle (setup, warm up, etc.).

File: mrp_operations_extension/models/mrp_routing.py [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    time_stop = fields.Float('Time after prod.',
+                             help="Time in hours for the cleaning.")
```
Comment (elicoidal) [inline mrp_operations_extension/models/mrp_routing.py]: Duration necessary after the cycle (cleaning, dismounting etc.).

File: mrp_operations_extension/models/mrp_routing.py [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+    op_number = fields.Integer('# operators', default='0')
+    op_avg_cost = fields.Float(
+        string='Operator avg. hour cost',
```
Comment (elicoidal) [inline mrp_operations_extension/models/mrp_routing.py]: s/avg./average
s/hour/hourly

File: mrp_operations_extension/models/mrp_workcenter.py [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+                                     string='Pre-operation costing product')
+    post_op_product = fields.Many2one('product.product',
+                                      string='Post-operation costing product')
+    rt_operations = fields.Many2many(
+        'mrp.routing.operation', 'mrp_operation_workcenter_rel', 'workcenter',
+        'operation', 'Routing Operations')
+    operators = fields.Many2many('res.users', 'mrp_wc_operator_rel',
+                                 'workcenter_id', 'operator_id', 'Operators')
+    op_number = fields.Integer(string='# Operators')
+    op_avg_cost = fields.Float(string='Operator average hour cost',
```
Comment (elicoidal) [inline mrp_operations_extension/models/mrp_workcenter.py]: s/hour/hourly

File: mrp_operations_extension/models/mrp_workcenter.py [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+    pre_op_product = fields.Many2one('product.product',
+                                     string='Pre-operation costing product')
+    post_op_product = fields.Many2one('product.product',
+                                      string='Post-operation costing product')
+    rt_operations = fields.Many2many(
+        'mrp.routing.operation', 'mrp_operation_workcenter_rel', 'workcenter',
+        'operation', 'Routing Operations')
+    operators = fields.Many2many('res.users', 'mrp_wc_operator_rel',
+                                 'workcenter_id', 'operator_id', 'Operators')
+    op_number = fields.Integer(string='# Operators')
```
Comment (elicoidal) [inline mrp_operations_extension/models/mrp_workcenter.py:35]: is it the operator's number or the number of operators?

File: mrp_operations_extension/models/res_config.py [hunk 15]
```diff
@@ -0,0 +1,13 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+# For copyright and license notices, see __openerp__.py file in root directory
+##############################################################################
+from openerp import models, fields
```
Comment (elicoidal) [inline mrp_operations_extension/models/res_config.py]: sort order

File: mrp_operations_extension/models/res_config.py [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+
+
+class MrpConfigSettings(models.TransientModel):
+    _inherit = 'mrp.config.settings'
+
+    group_mrp_workers = fields.Boolean(
+        string='Manage operators in work centers ',
```
Comment (elicoidal) [inline mrp_operations_extension/models/res_config.py]: ending space

File: mrp_operations_project/README.rst [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+
+==================================
+MRP Project Link (with operations)
+==================================
+
+This module links manufacturing operations with projects. It performs the
+following actions:
+
+* Auto-create tasks for work orders with assigned operators.
+* Add a tab where coding task works for the work order.
```
Comment (elicoidal) [inline mrp_operations_project/README.rst]: "where to encode task for the work order."

File: mrp_operations_project/README.rst [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+* Add a link on task for the scheduled products for that work order.
```
Comment (elicoidal) [inline mrp_operations_project/README.rst]: Add a link on task for the work order scheduled product.

File: mrp_operations_project/README.rst [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+Installation
+============
+
+This modules is auto-installed when you install *mrp_operations_extension* and
+*mrp_project*.
+
+Usage
+=====
+
+In a manufacturing order (MO) that has work orders, and they have at least
```
Comment (elicoidal) [inline mrp_operations_project/README.rst]: s/that has work orders, and they have at least/containing work orders with at least

File: mrp_operations_project/README.rst [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+one operator assigned, when you start that work order, a task is created with
```
Comment (elicoidal) [inline mrp_operations_project/README.rst]: when a work order is started, a task is created and assigned to the operator.

File: mrp_operations_project/README.rst [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+that operator as responsible of it.
+
+Navigating through each work order, there's a new tab called "Operators time"
```
Comment (elicoidal) [inline mrp_operations_project/README.rst]: In the work order form, a new tab "Operators time" is added to input the the work log.

NB: Operators time is misleading IMO. I would suggest "Work log"

File: mrp_operations_project/README.rst [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+where the task works can be put in an easy way.
+
+And last, if you go to a task linked to a work order, you'll see the scheduled
```
Comment (elicoidal) [inline mrp_operations_project/README.rst]: "Besides, in the related tasks, the products scheduled in the work order are listed."

**Comment by JavierIniesta** [comment]
Tested in local environment :+1:

**Comment by rafaelbn** [comment]
Runbot says "Killed"

**Comment by rafaelbn** [comment]
Tested in local enviroment :+1:  (runbot is very slowly today)

**Comment by pedrobaeza** [comment]
Rebased on top of last recent merge of mrp_operations_extension.

@rafaelbn, @JavierIniesta, @oihane, give it last check before merging.

File: mrp_operations_project/README.rst (hunk 21 above)
Comment (pedrobaeza) [inline mrp_operations_project/README.rst]: I prefer current name because there are machine and operators time tabs

**Comment by rafaelbn** [comment]
Hi @pedrobaeza , this is not a bug but a question: is it possible to create tasks before when confirming Manufacturing Order and not later when starting Work order?

**Comment by JMODaERP** [comment]
I tested, O.K.  :+1:

**Comment by JavierIniesta** [comment]
:+1:

**Comment by rafaelbn** [comment]
Discussed internaly, OK let leave creation of task que Work order starts. Then :+1:

**Comment by anajuaristi** [comment]
@rafaelbn That was a hot topic when we were designing module. Both aproaches would be right but at the end our customers decided to create the tasks at the moment where workorder starts, simply because if you decide to cancell the MO or even the WO before starting it, or just plan or change dates before starting, it was easier to manage integrity between WO and Tasks.

**Comment by pedrobaeza** [comment]
I merge for continuing the rest of the PRs as the module has been reviewed enough.

**Pull request merged**